org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/ExecutorServiceProvider.java

```diff
@@ -0,0 +1,68 @@
+/**
+ * Copyright (c) 2016, 2020 TypeFox GmbH (http://www.typefox.io) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.ide;
+
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.eclipse.xtext.util.DisposableRegistry;
+import org.eclipse.xtext.util.IDisposable;
+
+import com.google.common.collect.Maps;
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+import com.google.inject.Singleton;
+
+/**
+ * Provider for executor services. By calling {@link #dispose()} all created executor services are shut down.
+ * <p>
+ * In some situations it is necessary to use multiple instances of executor services in order to avoid deadlocks. That
+ * can be achieved with the {@link #get(String)}�method, which will return a different instance for each key.
+ */
+@Singleton
+public class ExecutorServiceProvider implements Provider<ExecutorService>, IDisposable {
+	@Inject
+	public void registerTo(DisposableRegistry disposableRegistry) {
+		disposableRegistry.register(this);
+	}
+
+	private final Map<String, ExecutorService> instanceCache = Maps.newHashMapWithExpectedSize(3);
```

Comment:
Should be a ConcurrentHashMap, otherwise the synchronization attempt in get(String) is completely flawed

Reply:
why? there is a second read in the synchronized

Reply:
Thread A calls get, sees nothing and enters the synchronize block, calls get again, sees still nothing and starts to put things into the map (this is not an atomic operation). While the map is being modified, Thread B calls get and sees an inconsistent map state. This can be avoided by using a ConcurrentHashMap + computeIfAbsent

Reply:
sry. can we do this in a separate PR. i am not able to think this properly through right now

Reply:
Agreed.